File: benchmarks/benchmark_rnn.py
"""
RNN (GRU) Training Benchmark: JAX-MLX
2-layer GRU → Dense(relu) → Dense(10) with SGD.
Matches the architecture of native MLX and Julia benchmarks.

Usage:
  python benchmarks/benchmark_rnn.py
"""

import os
if "MLX_PJRT_DEBUG" in os.environ:
    del os.environ["MLX_PJRT_DEBUG"]
if "JAX_PLATFORMS" not in os.environ:
    os.environ["JAX_PLATFORMS"] = "cpu,mlx"
elif "cpu" not in os.environ["JAX_PLATFORMS"]:
    os.environ["JAX_PLATFORMS"] = "cpu," + os.environ["JAX_PLATFORMS"]

import jax
import jax.numpy as jnp
from flax import nnx
import time
import numpy as np


class GRULayer(nnx.Module):
    """Single GRU layer using lax.scan with full unroll."""

    def __init__(self, input_dim: int, hidden_dim: int, *, rngs: nnx.Rngs):
        self.hidden_dim = hidden_dim
        self.input_proj = nnx.Linear(input_dim, 3 * hidden_dim, rngs=rngs)
        self.hidden_proj = nnx.Linear(hidden_dim, 3 * hidden_dim, use_bias=False, rngs=rngs)

    def __call__(self, x):
        """
        Args:
            x: (batch, seq_len, input_dim)
        Returns:
            h_final: (batch, hidden_dim) — final hidden state
            h_all:   (batch, seq_len, hidden_dim) — all hidden states
        """
        batch, seq_len, _ = x.shape
        H = self.hidden_dim
        h0 = jnp.zeros((batch, H))

        # Pre-compute input projections for all timesteps
        x_proj = jax.vmap(self.input_proj)(x.reshape(-1, x.shape[-1])).reshape(batch, seq_len, 3 * H)
        x_proj = jnp.transpose(x_proj, (1, 0, 2))  # (seq_len, batch, 3*H)

        hidden_proj = self.hidden_proj

        def gru_step(h, x_t):
            hx = hidden_proj(h)
            z = jax.nn.sigmoid(x_t[:, :H] + hx[:, :H])
            r = jax.nn.sigmoid(x_t[:, H:2*H] + hx[:, H:2*H])
            n = jnp.tanh(x_t[:, 2*H:] + r * hx[:, 2*H:])
            h_new = (1 - z) * n + z * h
            return h_new, h_new

        h_final, h_all = jax.lax.scan(gru_step, h0, x_proj, unroll=seq_len)
        # h_all: (seq_len, batch, H) → (batch, seq_len, H)
        h_all = jnp.transpose(h_all, (1, 0, 2))
        return h_final, h_all


class GRUClassifier(nnx.Module):
    """2-layer GRU → Dense(relu) → Dense(10). Matches Julia/native."""

    def __init__(self, input_dim: int, hidden_dim: int, num_classes: int, *, rngs: nnx.Rngs):
        self.gru1 = GRULayer(input_dim, hidden_dim, rngs=rngs)
        self.gru2 = GRULayer(hidden_dim, hidden_dim, rngs=rngs)
        self.dense1 = nnx.Linear(hidden_dim, hidden_dim, rngs=rngs)
        self.dense2 = nnx.Linear(hidden_dim, num_classes, rngs=rngs)

    def __call__(self, x):
        _, h1_all = self.gru1(x)         # h1_all: (batch, seq_len, hidden)
        h2_final, _ = self.gru2(h1_all)  # h2_final: (batch, hidden)
        h = jax.nn.relu(self.dense1(h2_final))
        return self.dense2(h)


def cross_entropy_loss(model, x, y):
    logits = model(x)
    one_hot = jax.nn.one_hot(y, logits.shape[-1])
    return -jnp.mean(jnp.sum(jax.nn.log_softmax(logits) * one_hot, axis=-1))


import optax


@nnx.jit
def train_step(model, optimizer, x, y):
    loss, grads = nnx.value_and_grad(cross_entropy_loss)(model, x, y)
    optimizer.update(model, grads)
    return loss


def benchmark_rnn(batch_size=128, seq_len=64, input_dim=64, hidden_dim=512,
                  num_classes=10, num_warmup=30, num_runs=20):
    """2-layer GRU training benchmark."""

    try:
        device = jax.devices("mlx")[0]
        device_name = "mlx"
    except Exception:
        device = jax.devices("cpu")[0]
        device_name = "cpu"

    print(f"Device: {device_name}")
    print(f"Config: batch={batch_size}, seq_len={seq_len}, input_dim={input_dim}, "
          f"hidden={hidden_dim}, classes={num_classes}")
    print(f"Model: 2× GRU({input_dim}→{hidden_dim}, {seq_len} steps) → Dense({hidden_dim}, relu) → Dense({num_classes})")
    print(f"Task: Forward + Backward + SGD\n")

    lr = 0.001

    with jax.default_device(device):
        rngs = nnx.Rngs(0)
        model = GRUClassifier(input_dim, hidden_dim, num_classes, rngs=rngs)
        optimizer = nnx.Optimizer(model, optax.sgd(lr), wrt=nnx.Param)

        key = jax.random.PRNGKey(42)
        k1, k2 = jax.random.split(key)
        x = jax.random.normal(k1, (batch_size, seq_len, input_dim))
        y = jax.random.randint(k2, (batch_size,), 0, num_classes)

        # Warmup
        print(f"Warming up ({num_warmup} steps)...")
        for _ in range(num_warmup):
            loss = train_step(model, optimizer, x, y)
            loss.block_until_ready()
        print("Warmup done ✓\n")

        # Timed runs
        print(f"{'='*50}")
        print("Timed runs")
        print(f"{'='*50}")
        times = []
        for i in range(num_runs):
            t0 = time.perf_counter()
            loss = train_step(model, optimizer, x, y)
            loss.block_until_ready()
            elapsed = (time.perf_counter() - t0) * 1000
            times.append(elapsed)
            print(f"  Run {i+1}: {elapsed:.2f}ms (loss: {float(loss):.4f})")

        mean_t = np.mean(times)
        std_t = np.std(times)
        min_t = np.min(times)
        print(f"\n  Mean: {mean_t:.2f}ms ± {std_t:.2f}ms  Min: {min_t:.2f}ms")

    print(f"\n{'='*50}")
    print("SUMMARY")
    print(f"{'='*50}")
    print(f"JAX-MLX: {mean_t:.2f}ms ± {std_t:.2f}ms")
    return mean_t, std_t


if __name__ == "__main__":
    print("RNN (GRU) Training Benchmark — JAX-MLX")
    print("Architecture: 2-layer GRU → Dense(relu) → Dense(10)\n")
    benchmark_rnn()

File: benchmarks/benchmark_rnn.jl
"""
RNN (GRU) Training Benchmark (Julia)
2-layer GRU → Dense(relu) → Dense(10) with SGD.
Uses GRUCell with manual unrolling. Matches JAX-MLX and native MLX architectures.

Usage:
  julia +1.11 benchmarks/benchmark_rnn.jl
"""

using Libdl, Reactant, Lux, Enzyme, Random, Statistics

# ── MLX Plugin Setup ──────────────────────────────────────────────────────────

function load_mlx_plugin!()
    mlx_lib = "/Users/thomas/miniforge3/envs/jax/lib/python3.13/site-packages/mlx/lib/libmlx.dylib"
    Libdl.dlopen(mlx_lib, Libdl.RTLD_NOW | Libdl.RTLD_GLOBAL)

    plugin_path = joinpath(@__DIR__, "..", "src", "jax_mlx", "libmlx_pjrt_plugin.dylib")
    if !isfile(plugin_path)
        plugin_path = joinpath(@__DIR__, "..", "build", "libmlx_pjrt_plugin.dylib")
    end
    @assert isfile(plugin_path) "Cannot find libmlx_pjrt_plugin.dylib at $plugin_path"

    mlx_client_ptr = Reactant.XLA.PJRT.MakeClientUsingPluginAPI(plugin_path, "mlx", "MLX")
    mlx_client = Reactant.XLA.PJRT.Client(mlx_client_ptr)
    Reactant.XLA.global_backend_state.clients["mlx"] = mlx_client
    Reactant.XLA.set_default_backend(mlx_client)
    println("MLX plugin loaded ✓")
end

# ── 2-Layer Manual GRU ────────────────────────────────────────────────────────
# Manually unrolls two GRUCells over timesteps.
# Reactant traces through the for loops → straight-line compute (no while_loop).

struct TwoLayerGRU{S,C1,C2,D1,D2} <: Lux.AbstractLuxContainerLayer{(:gru_cell1, :gru_cell2, :dense1, :dense2)}
    seq_len::S
    gru_cell1::C1
    gru_cell2::C2
    dense1::D1
    dense2::D2
end

function TwoLayerGRU(; input_dim=64, hidden_dim=512, seq_len=64, num_classes=10)
    return TwoLayerGRU(
        seq_len,
        GRUCell(input_dim => hidden_dim),
        GRUCell(hidden_dim => hidden_dim),
        Dense(hidden_dim, hidden_dim, relu),
        Dense(hidden_dim, num_classes),
    )
end

function (m::TwoLayerGRU)(x, ps, st)
    # x: (input_dim, seq_len, batch)
    batch_size = size(x, 3)

    # --- Layer 1: GRU over input ---
    x_t = x[:, 1, :]  # (input_dim, batch)
    (h1, (h1_state,)), st_gru1 = m.gru_cell1(x_t, ps.gru_cell1, st.gru_cell1)

    # Collect all hidden states for layer 2
    h1_all = similar(h1, size(h1, 1), m.seq_len, batch_size)
    h1_all[:, 1, :] = h1

    for t in 2:m.seq_len
        x_t = x[:, t, :]
        (h1, (h1_state,)), st_gru1 = m.gru_cell1((x_t, (h1_state,)), ps.gru_cell1, st_gru1)
        h1_all[:, t, :] = h1
    end

    # --- Layer 2: GRU over layer 1's hidden states ---
    x2_t = h1_all[:, 1, :]
    (h2, (h2_state,)), st_gru2 = m.gru_cell2(x2_t, ps.gru_cell2, st.gru_cell2)

    for t in 2:m.seq_len
        x2_t = h1_all[:, t, :]
        (h2, (h2_state,)), st_gru2 = m.gru_cell2((x2_t, (h2_state,)), ps.gru_cell2, st_gru2)
    end

    # Classification head on final hidden state
    y, st_d1 = m.dense1(h2, ps.dense1, st.dense1)
    y, st_d2 = m.dense2(y, ps.dense2, st.dense2)
    return y, (gru_cell1=st_gru1, gru_cell2=st_gru2, dense1=st_d1, dense2=st_d2)
end

# ── Loss Function ─────────────────────────────────────────────────────────────

function cross_entropy_loss(model, ps, st, x, y)
    logits, st_ = model(x, ps, st)
    log_probs = logits .- Lux.logsumexp(logits; dims=1)
    classes = reshape(1:10, 10, 1)
    labels = reshape(y, 1, :)
    one_hot = Float32.(classes .== labels)
    return -sum(one_hot .* log_probs) / Float32(size(logits, 2))
end

# ── Train Step ────────────────────────────────────────────────────────────────

function train_step(model, ps, st, x, y, lr)
    dps = Enzyme.make_zero(ps)
    _, loss = Enzyme.autodiff(
        Enzyme.ReverseWithPrimal,
        cross_entropy_loss,
        Active,
        Const(model),
        Duplicated(ps, dps),
        Const(st),
        Const(x),
        Const(y),
    )
    ps_new = Lux.recursive_map((p, g) -> p .- lr .* g, ps, dps)
    return ps_new, loss
end

# ── Benchmark ─────────────────────────────────────────────────────────────────

function benchmark_rnn(;
    batch_size::Int=128,
    seq_len::Int=64,
    input_dim::Int=64,
    hidden_dim::Int=512,
    num_warmup::Int=30,
    num_runs::Int=20,
    lr::Float32=0.001f0,
)
    println("RNN (GRU) Training Benchmark (Julia/Lux/Reactant)")
    println("Config: batch_size=$batch_size, seq_len=$seq_len, input_dim=$input_dim, hidden=$hidden_dim")
    println("Model: 2× GRU($input_dim→$hidden_dim, $seq_len steps, unrolled) → Dense($hidden_dim, relu) → Dense(10)")
    println("Task: Forward + Backward pass with SGD\n")

    load_mlx_plugin!()

    model = TwoLayerGRU(input_dim=input_dim, hidden_dim=hidden_dim, seq_len=seq_len)
    rng = Random.Xoshiro(0)
    ps, st = Lux.setup(rng, model)
    ps = ps |> Lux.f32

    # Lux format: (input_dim, seq_len, batch)
    x = randn(Float32, input_dim, seq_len, batch_size)
    y = rand(1:10, batch_size)

    x_r = Reactant.to_rarray(x)
    y_r = Reactant.to_rarray(y)
    ps_r = Reactant.to_rarray(ps)
    st_r = Reactant.to_rarray(st)

    println("Compiling train step...")
    t_compile_start = time()
    compiled_step = @compile train_step(model, ps_r, st_r, x_r, y_r, lr)
    t_compile = time() - t_compile_start
    println("Compilation done in $(round(t_compile; digits=1))s ✓\n")

    println("Warming up ($num_warmup steps)...")
    for i in 1:num_warmup
        ps_r, loss = compiled_step(model, ps_r, st_r, x_r, y_r, lr)
    end
    println("Warmup done ✓\n")

    println("="^50)
    println("Timed runs")
    println("="^50)
    times = Float64[]
    for i in 1:num_runs
        t0 = time_ns()
        ps_r, loss = compiled_step(model, ps_r, st_r, x_r, y_r, lr)
        loss_val = Float64(loss)
        t1 = time_ns()
        dt_ms = (t1 - t0) / 1e6
        push!(times, dt_ms)
        println("  Run $i: $(round(dt_ms; digits=2))ms (loss: $(round(loss_val; digits=4)))")
    end

    steady = times[2:end]
    mean_t = mean(steady)
    std_t = std(steady)
    min_t = minimum(steady)
    println("\n  Mean: $(round(mean_t; digits=2))ms ± $(round(std_t; digits=2))ms  Min: $(round(min_t; digits=2))ms  (excl. first run)")

    println("\n" * "="^50)
    println("SUMMARY")
    println("="^50)
    println("Julia/Lux/Reactant/MLX: $(round(mean_t; digits=2))ms ± $(round(std_t; digits=2))ms")
end

benchmark_rnn()
